File: cpp/KalmanFilter.cpp
#include <iostream>
#include "KalmanFilter.hpp"


int main(int argc, char* argv[]){
    LKF linear_kalman_filter(87.16960907, 133.9387207);

    Eigen::Vector2f u_k;
    float acc_x_k = 2.344874339221101;
    float acc_y_k = 1.9954219601903327;
    float vel_x_k = 0.3214601508381209;
    float vel_y_k = 0.24648556635222357;

    float theta_t = 0.10085797309875488;
    u_k <<  vel_x_k + 0.5f * acc_x_k * theta_t, vel_y_k + 0.5f * acc_y_k * theta_t;

    Eigen::Vector2f z_k;
    z_k << 87.1461181640625, 133.94679260253906;

    Eigen::Vector2f output = linear_kalman_filter.run(0.10085797309875488, u_k, z_k);

    std::cout<<output<<std::endl;

    return 0;
}

int test(){

    //// Unchanged variables

    Eigen::Matrix2f H_k; // Scale Matrix
    H_k <<      1.0f,       0.0f,
                0.0f,       1.0f;

    Eigen::Matrix2f Q_k; // Prediction Noise
    Q_k <<      1.0f,       0.0f,
                0.0f,       1.0f;
    Eigen::Matrix2f R_k; // Observation Noise
    R_k <<      1.0f,       0.0f,
                0.0f,       1.0f;


    //// Updated variables

    Eigen::Vector2f x_hat_k_prev; // Previous Value
    x_hat_k_prev <<     87.16960907,
                        133.9387207;

    Eigen::Matrix2f P_k_prev; // Previous prediction covariance matrix
    P_k_prev <<     1.0f,       0.0f,
                    0.0f,       1.0f;

    float theta_t = 0.10085797309875488; // *

    // Prediction
    Eigen::Matrix2f F_k; // Prediction Matrix
    F_k <<      1.0f,       0.0f,
                0.0f,       1.0f;
    
    Eigen::Matrix2f B_k;
    B_k <<      theta_t,    0.0f,
                0.0,        theta_t;

    float acc_x_k = 2.344874339221101;
    float acc_y_k = 1.9954219601903327;
    float vel_x_k = 0.3214601508381209;
    float vel_y_k = 0.24648556635222357;

    Eigen::Vector2f u_k; // Control Vector *
    u_k <<  vel_x_k + 0.5f * acc_x_k * theta_t, vel_y_k + 0.5f * acc_y_k * theta_t;

    Eigen::Vector2f x_hat_k; // Prediction Result
    x_hat_k = F_k * x_hat_k_prev + B_k * u_k;

    Eigen::Matrix2f P_k; // Prediction Covariance
    P_k = F_k * P_k_prev * F_k.transpose() + Q_k;
    
    // Update
    Eigen::Vector2f z_k; // Observation Vector *
    z_k << 87.1461181640625, 133.94679260253906, -0.232910746, 0.0800323382;

    Eigen::Matrix2f K_prime; // Kalman gain(Modified)
    K_prime = P_k * H_k * (H_k * P_k * H_k.transpose() + R_k).completeOrthogonalDecomposition().pseudoInverse();
    
    Eigen::Vector2f x_hat_prime_k; // Update result
    x_hat_prime_k = x_hat_k + K_prime * (z_k - H_k * x_hat_k);

    Eigen::Matrix2f P_prime_k; // Update prediction covariance
    P_prime_k = P_k - K_prime * H_k * P_k;

    return 0;
}

LKF::LKF(float init_pose_x, float init_pose_y){
    H_k <<      1.0f,       0.0f,
                0.0f,       1.0f;

    Q_k <<      1.0f,       0.0f,
                0.0f,       1.0f;

    R_k <<      1.0f,       0.0f,
                0.0f,       1.0f;

    x_hat_k_prev << init_pose_x,    init_pose_y;

    P_k_prev << 1.0f,       0.0f,
                0.0f,       1.0f;
}

Eigen::Vector2f LKF::run(float theta_t, Eigen::Vector2f u_k, Eigen::Vector2f z_k){ // z_k: observation vector
    // Prediction
    Eigen::Matrix2f F_k; // Prediction Matrix
    F_k <<      1.0f,       0.0f,
                0.0f,       1.0f;

    Eigen::Matrix2f B_k; // Control Matrix
    B_k <<      theta_t,    0.0f,
                0.0,        theta_t;

    Eigen::Vector2f x_hat_k; // Predict Result
    x_hat_k = F_k * x_hat_k_prev + B_k * u_k;

    Eigen::Matrix2f P_k; // Prediction Covariance
    P_k = F_k * P_k_prev * F_k.transpose() + Q_k;

    // Update
    Eigen::Matrix2f K_prime; // Kalman gain(Modified)
    K_prime = P_k * H_k * (H_k * P_k * H_k.transpose() + R_k).completeOrthogonalDecomposition().pseudoInverse();

    Eigen::Vector2f x_hat_prime_k; // Update result
    x_hat_prime_k = x_hat_k + K_prime * (z_k - H_k * x_hat_k);

    Eigen::Matrix2f P_prime_k; // Update prediction covariance
    P_prime_k = P_k - K_prime * H_k * P_k;

    x_hat_k_prev = x_hat_prime_k;
    P_k_prev = P_prime_k;

    return x_hat_prime_k;
}



/*
[[ 87.16960907]     // pose_x
 [133.9387207 ]     // pose_y
 [  0.32146015]     // vel_x
 [  0.24648557]]    // vel_y
 */

File: scripts/KalmanFilter.py
import rosbag
from geometry_msgs.msg import PoseStamped
import matplotlib.pyplot as plt
import numpy as np
import copy
import random

def create_inputs(data):
    inputs = []
    for i in range(len(data[0])):
        inputs.append([data[0][i], data[1][i], data[2][i], data[3][i]])

    return inputs 

def main():
    bag = rosbag.Bag('../files/220203_FMTC_ndt_pose_with_gps_data_red_course.bag')

    ndt_noise_flag = False # Add noise to ndt pose data
    filter_observation_noise_flag = False # When noise is exist, use prediction value for observation

    pose_list = []
    time_list = []
    noise_occurence = []

    for topic, msg, t in bag.read_messages(topics=['/ndt_pose']):
        pose_list.append([msg.pose.position.x, msg.pose.position.y, msg.pose.position.z, msg.pose.orientation.x, msg.pose.orientation.y, msg.pose.orientation.z, msg.pose.orientation.w])
        time_list.append(msg.header.stamp.secs+msg.header.stamp.nsecs/1000000000)

    vel_x_list = []
    vel_y_list = []

    for i in range(1,len(pose_list)):
        vel_x = (pose_list[i][0] - pose_list[i-1][0])/(time_list[i] - time_list[i-1])
        vel_y = (pose_list[i][1] - pose_list[i-1][1])/(time_list[i] - time_list[i-1])
        vel_x_list.append(vel_x)
        vel_y_list.append(vel_y)


    acc_x_list = []
    acc_y_list = []

    for i in range(1, len(vel_x_list)):
        acc_x = (vel_x_list[i] - vel_x_list[i-1])/(time_list[i+1] - time_list[i])
        acc_y = (vel_y_list[i] - vel_y_list[i-1])/(time_list[i+1] - time_list[i])
        acc_x_list.append(acc_x)
        acc_y_list.append(acc_y)

    cnt = 0
    for i in range(len(pose_list)):
        x = pose_list[i][0]
        y = pose_list[i][1]

        noise_occurence.append(0)

        if(ndt_noise_flag and random.randrange(1,200) == 1):
            noise = np.random.normal(0, 15, 2)
            x = x + noise[0]
            y = y + noise[1]
            pose_list[i][0] = x
            pose_list[i][1] = y
            noise_occurence[i] = 1
                

    
    # Align index of data structures
    pose_x_list = [pose[0] for pose in pose_list]
    pose_x_list = pose_x_list[2:]
    pose_y_list = [pose[1] for pose in pose_list]
    pose_y_list = pose_y_list[2:]
    vel_x_list = vel_x_list[1:]
    vel_y_list = vel_y_list[1:]
    time_list = time_list[2:]
    noise_occurence = noise_occurence[2:]

    # initialization
    pose_x = pose_x_list[0]
    pose_y = pose_y_list[0]
    vel_x = vel_x_list[0]
    vel_y = vel_y_list[0]

    x_hat_prime_k = np.array([[pose_x], [pose_y]])
    P_prime_k = np.identity(2) # Initialize covariance as Identical matrix

    # Scale Matrix
    H_k = np.array([[1,      0],
                    [0,      1]])

    # Prediction Noise
    Q_I = 1
    Q_k = np.array([[Q_I,      0],
                    [0,      Q_I]])
    
    # Observation Noise
    R_I = 1
    R_k = np.array([[R_I,      0],
                    [0,      R_I]])

    # Output
    output_pose_x_list = []
    output_pose_y_list = []

    for i in range(1, len(pose_x_list)):
        x_hat_k_prev = copy.deepcopy(x_hat_prime_k)
        P_k_prev = copy.deepcopy(P_prime_k)
        theta_t = time_list[i] - time_list[i-1]

        # Predict
        F_k = np.array([[1,     0],
                        [0,     1]])
        
        B_k = np.array([[theta_t,       0         ], 
                        [0,             theta_t   ]])


        vel_x_k = vel_x_list[i-1]
        vel_y_k = vel_y_list[i-1]
        acc_x_k = acc_x_list[i-1]
        acc_y_k = acc_y_list[i-1]

        u_k = np.array([[vel_x_k + 0.5 * acc_x_k * theta_t], [vel_y_k + 0.5 * acc_y_k * theta_t]])

        x_hat_k = F_k @ x_hat_k_prev + B_k @ u_k
        P_k = F_k @ P_k_prev @ np.transpose(F_k) + Q_k

        # Update
        pose_x = pose_x_list[i]
        pose_y = pose_y_list[i]

        if(filter_observation_noise_flag and noise_occurence[i] == 1):
            pose_x = x_hat_k[0]
            pose_y = x_hat_k[1]

        z_k = np.array([[pose_x], [pose_y]])

        K_prime = P_k @ H_k @ np.linalg.pinv(H_k @ P_k @ np.transpose(H_k) + R_k)

        x_hat_prime_k = x_hat_k + K_prime @ (z_k - H_k @ x_hat_k)
        P_prime_k = P_k - K_prime @ H_k @ P_k

        output_pose_x_list.append(x_hat_prime_k[0])
        output_pose_y_list.append(x_hat_prime_k[1])

    plt.plot(pose_x_list, pose_y_list, '-p', markersize=1, linewidth=0.5, label='origin')
    plt.plot(output_pose_x_list, output_pose_y_list, '-p', markersize=1, linewidth=0.5, label='KF')
    plt.legend()
    plt.grid(True)


    plt.show()
    plt.close()


    return

if __name__ == '__main__':
    main()

'''
header: 
  seq: 10449
  stamp: 
    secs: 1643870365
    nsecs: 239412000
  frame_id: "/map"
pose: 
  position: 
    x: 87.12862396240234
    y: 133.9093017578125
    z: 6.892755031585693
  orientation: 
    x: 0.06568801780013588
    y: 0.0318829475407855
    z: -0.00801276376061509
    w: 0.9972985298247045

'''